dll/win32/kernel32/kernel32_vista/vista.c

```diff
@@ -62,18 +117,69 @@ QueryFullProcessImageNameW(HANDLE hProcess,
 
     if (!NT_SUCCESS(Status)) goto Cleanup;
 
-    if (Result->Length / sizeof(WCHAR) + 1 > *pdwSize)
+    /* The name should use the Win32 path format. */
+    if (dwFlags == 0)
     {
-        Status = STATUS_BUFFER_TOO_SMALL;
-        goto Cleanup;
+        pszResult = RtlAllocateHeap(RtlGetProcessHeap(), 0, Result->Length + sizeof(WCHAR));
+
+        if (!pszResult)
+        {
+            Status = STATUS_NO_MEMORY;
+            goto Cleanup;
+        }
+
+        RtlCopyMemory(pszResult, Result->Buffer, Result->Length);
+        pszResult[Result->Length / sizeof(WCHAR)] = UNICODE_NULL;
+
+        if (!NormalizeDevicePath(pszResult, (Result->Length / sizeof(WCHAR)) + 1))
+        {
+            Status = STATUS_UNSUCCESSFUL;
+            goto Cleanup;
+        }
+
+        dwLength = wcslen(pszResult);
+
+        if (dwLength + 1 > *pdwSize)
+        {
+            Status = STATUS_BUFFER_TOO_SMALL;
+            goto Cleanup;
+        }
+
+        StringCchCopyW(lpExeName, *pdwSize, pszResult);
+
+        *pdwSize = dwLength;
     }
+    /* The name should use the native system path format. */
+    else if (dwFlags == PROCESS_NAME_NATIVE)
+    {
+        dwLength = Result->Length / sizeof(WCHAR);
+
+        if (dwLength + 1 > *pdwSize)
+        {
+            Status = STATUS_BUFFER_TOO_SMALL;
+            goto Cleanup;
+        }
 
-    *pdwSize = Result->Length / sizeof(WCHAR);
-    memcpy(lpExeName, Result->Buffer, Result->Length);
-    lpExeName[*pdwSize] = 0;
+        RtlCopyMemory(lpExeName, Result->Buffer, Result->Length);
+        lpExeName[dwLength] = UNICODE_NULL;
```

Comment:
Ditto

Reply:
For null-terminating, I use as suggested in [https://reactos.org/wiki/Coding_Style](https://reactos.org/wiki/Coding_Style)

Reply:
What I meant is to use something like:
```c
RtlStringCbCopyW(lpExeName, Result->Buffer, Result->Length);
```

Reply:
I used RtlStringCchCopyUnicodeString instead, fixed.

Reply:
The length has already been validated here, so RtlStringCbCopyNW seems like overkill.